core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala

```diff
@@ -801,12 +803,130 @@ class CompactEntityQuery(driverComponent: DriverComponent)
+          and ("""
+
+    concatSqlActions(baseSql, clause, sql"))")
       .as[Boolean]
       .head
+  }
+
+  /**
+    * Removed the specified attributes from all entities of the given type and workspace.
+    *
+    * `execution plan: index range scan on idx_entity_type_name (with lots of JSON and string manipulation)`
+    */
+  def deleteAttributes(workspaceId: UUID,
+                       entityType: String,
+                       attributeNames: Set[AttributeName]
+  ): ReadWriteAction[Int] =
+
+    if (attributeNames.isEmpty) {
+      DBIO.failed(
+        new AttributeException(
+          message = "The supplied set of attributes to remove cannot be empty.",
+          code = StatusCodes.BadRequest
+        )
+      )
+    } else {
+      // Map all attributes-to-be-removed into a single regex. In pseudocode, this regex looks for:
+      //     {"a": "attr1 OR attr2 OR attr3", ... },
+      // with the final comma being optional.
+      // This should hit on any instance of any element of `attributeNames` inside the $.refs array.
+      val allAttrNames = attributeNames.map(x => AttributeName.toDelimitedName(x).replace(":", "\\:")).mkString("|")
+      val regex = s"""\\{\\"a\\": \\"(?:$allAttrNames)\\",[^}]+\\},?"""
+
+      // Remove all elements in the $.refs array for the attributes we want to delete.
+      // This is done via JSON_REPLACE(CAST(REPLACE(REGEXP_REPLACE(JSON_EXTRACT)))). Explaining from the inside out:
+      //   - JSON_EXTRACT(attributes, '$$.refs') gets the refs array
+      //   - REGEXP_REPLACE(...) treats the refs array as a plain string, and deletes all elements matching our regex
+      //   - REPLACE(...) handles the case where we have removed the last object in the $.refs array
+      //                  and therefore need to remove the trailing comma
+      //   - CAST(... as JSON) converts the modified string back to a JSON array
+      //   - JSON_REPLACE(...) replaces the original refs array with the modified one
+      // We use a string replace here because it is significantly more performant than calling JSON_REMOVE
+      //  individually for each value that needs to be changed. Since we control the serialization format of the
+      //  refs array, and only perform the string replace inside that array, we avoid any problems with other
+      //  user-supplied values.
+      //
+      // All of this will be wrapped by JSON_REMOVE later - see the `removeSql` variable. That wrapping JSON_REMOVE
+      //  will handle removing the attributes from the `attrs` object.
+      val replaceRefsSql = sql"""JSON_REPLACE(attributes,
+                                              $slickRefsPath,
+                                              CAST(
+                                              REPLACE(
+                                                REGEXP_REPLACE(JSON_EXTRACT(attributes, $slickRefsPath), $regex, ''),
+                                                ', ]',
```

Comment:
you sure there's always going to be a space after a dangling comma?

Reply:
In practice, that's what I was seeing. This is only an issue when we delete the last element in the array. Consider a request to delete attribute "third":
```
[{"a": "first" ...}, {"a": "second" ...}, {"a": "third" ...}]
```
becomes
```
[{"a": "first" ...}, {"a": "second" ...}, ]
```
and so there's a trailing comma. But I agree this is fragile; I'll change it to another regex with flexible whitespace.

Reply:
which I now realize would fail if we deleted the last *two* elements in the array:
```
[{"a": "first" ...}, {"a": "second" ...}, {"a": "third" ...}]
```
becomes
```
[{"a": "first" ...},  ]
```
which has two trailing spaces